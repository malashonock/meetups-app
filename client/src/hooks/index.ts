export * from './useMeetupQuery';
export * from './useNewsArticleQuery';
export * from './useStaticFileQuery';
export * from './useAuthStore';
