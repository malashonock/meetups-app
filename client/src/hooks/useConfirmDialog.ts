import { useContext } from 'react';

import { ConfirmDialogContext, ConfirmFn } from 'components';
import { Optional } from 'types';

export const useConfirmDialog = (): ConfirmFn => {
  return useContext(ConfirmDialogContext);
};
