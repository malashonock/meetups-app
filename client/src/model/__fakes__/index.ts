export * from './user.fake';
export * from './meetup.fake';
export * from './file.fake';
