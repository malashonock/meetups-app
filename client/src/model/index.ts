export * from './meetup.model';
export * from './news.model';
export * from './user.model';
export * from './login.model';
