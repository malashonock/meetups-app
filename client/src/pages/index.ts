export * from './LoginPage/LoginPage';
export * from './MeetupsPage/MeetupsPage';
export * from './ViewMeetupPage/ViewMeetupPage';
export * from './CreateMeetupPage/CreateMeetupPage';
export * from './EditMeetupPage/EditMeetupPage';
export * from './NewsPage/NewsPage';
export * from './ViewNewsPage/ViewNewsPage';
export * from './CreateNewsPage/CreateNewsPage';
export * from './EditNewsPage/EditNewsPage';
export * from './NotFoundPage/NotFoundPage';
