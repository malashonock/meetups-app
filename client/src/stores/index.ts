export * from './root.store';
export * from './auth.store';
export * from './user.store';
export * from './ui.store';
export * from './meetup.store';
export * from './news.store';
