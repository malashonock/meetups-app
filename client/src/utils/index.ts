export * from './string';
export * from './date';
export * from './number';
export * from './file';
export * from './array';
export * from './test';
export * from './dom';
