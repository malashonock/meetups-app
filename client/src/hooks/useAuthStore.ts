import { useContext } from 'react';

import { AuthStore } from 'stores';
import { RootContext } from 'components';

export const useAuthStore = (): AuthStore => {
  return useContext(RootContext).authStore;
};
