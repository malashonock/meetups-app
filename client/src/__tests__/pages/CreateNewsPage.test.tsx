/* eslint-disable testing-library/no-unnecessary-act */

import { PropsWithChildren } from 'react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { act, fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { CreateNewsPage } from 'pages';
import { NewsFields } from 'model';
import { mockImageWithUrl } from 'model/__fakes__';
import { dropFile } from 'utils';
import { useNewsStore } from 'hooks';
import { RootStore } from 'stores';

jest.mock('utils/file');

// Mock useNewsStore hook
jest.mock('hooks', () => {
  return {
    ...jest.requireActual('hooks'),
    useNewsStore: jest.fn(),
  };
});
const mockUseNewsStore = useNewsStore as jest.MockedFunction<
  typeof useNewsStore
>;
const mockNewsStoreCreateArticle = jest.fn();

beforeEach(() => {
  mockUseNewsStore.mockReturnValue({
    newsStore: {
      rootStore: new RootStore(),
      news: [],
      loadNews: jest.fn(),
      createNewsArticle: mockNewsStoreCreateArticle,
      findNewsArticle: jest.fn(),
      onNewsArticleDeleted: jest.fn(),
    },
  });
});

afterEach(() => {
  jest.resetAllMocks();
});

const MockRouter = ({ children }: PropsWithChildren): JSX.Element => (
  <MemoryRouter initialEntries={['/news', '/news/create']}>
    <Routes>
      <Route path="/news">
        <Route index element={<h1>News page</h1>} />
        <Route path="create" element={children} />
      </Route>
    </Routes>
  </MemoryRouter>
);

const getTitleInput = () =>
  screen.getByLabelText('formFields.news.title.label');
const getTextInput = () => screen.getByLabelText('formFields.news.text.label');
const getImageInput = () =>
  screen.getByLabelText('formFields.news.image.label') as HTMLInputElement;
const getImageDropbox = () => screen.getByTestId('image-dropbox');
const getImagePreview = () => screen.getByTestId('image-preview');
const getImage = () =>
  screen.getByAltText('imagePreview.imgAlt') as HTMLImageElement;
const getSubmitBtn = () => screen.getByText('formButtons.create');
const getBackBtn = () => screen.getByText('formButtons.back');

const mockNews: NewsFields = {
  title: 'News title',
  text: 'News text',
  image: mockImageWithUrl,
};

describe('CreateNewsPage', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<CreateNewsPage />, { wrapper: MockRouter });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should open a blank form', () => {
    render(<CreateNewsPage />, { wrapper: MockRouter });

    expect(getTitleInput()).toHaveValue('');
    expect(getTextInput()).toHaveValue('');
    expect(getImageInput().files?.length).toBe(0);
  });

  it('should accept user input', async () => {
    render(<CreateNewsPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.type(getTitleInput(), mockNews.title);
    });
    expect(getTitleInput()).toHaveValue(mockNews.title);

    await act(() => {
      userEvent.type(getTextInput(), mockNews.text);
    });
    expect(getTextInput()).toHaveValue(mockNews.text);

    await act(() => {
      dropFile(getImageDropbox(), mockImageWithUrl);
    });
    expect(getImagePreview()).toBeInTheDocument();
    expect(getImage().src).toBe(mockImageWithUrl.url);
  });

  it('should validate form fields', async () => {
    render(<CreateNewsPage />, { wrapper: MockRouter });

    expect(getSubmitBtn()).toBeDisabled();

    await act(() => {
      fireEvent.blur(getTitleInput());
    });
    expect(
      screen.getByText('formFields.news.title.errorText'),
    ).toBeInTheDocument();

    await act(() => {
      fireEvent.blur(getTextInput());
    });
    expect(
      screen.getByText('formFields.news.text.errorText'),
    ).toBeInTheDocument();
  });

  it('should handle form submit', async () => {
    render(<CreateNewsPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.type(getTitleInput(), mockNews.title);
      userEvent.type(getTextInput(), mockNews.text);
      dropFile(getImageDropbox(), mockImageWithUrl);
    });

    await act(() => {
      userEvent.click(getSubmitBtn());
    });

    expect(mockNewsStoreCreateArticle).toHaveBeenCalledWith(mockNews);
    expect(screen.getByText('News page')).toBeInTheDocument();
  });

  it('should be able to navigate back to news page', async () => {
    render(<CreateNewsPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.click(getBackBtn());
    });

    expect(screen.getByText('News page')).toBeInTheDocument();
  });
});
