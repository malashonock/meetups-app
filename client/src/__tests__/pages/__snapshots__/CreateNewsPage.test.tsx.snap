// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewsPage should match snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <section
      class="container"
    >
      <h1
        class="typography h1 heading"
      >
        createNewsPage.title
      </h1>
      <div
        class="contentWrapper"
      >
        <div
          class="textSection main"
        >
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="title"
            >
              formFields.news.title.label
            </label>
            <input
              class="input default"
              data-testid="text-input"
              id="title"
              name="title"
              type="text"
              value=""
            />
          </div>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="text"
            >
              formFields.news.text.label
            </label>
            <div
              class="container"
            >
              <textarea
                class="input default textArea"
                data-testid="text-area"
                id="text"
                name="text"
                rows="1"
                style="height: 2px;"
              />
            </div>
          </div>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="image"
            >
              formFields.news.image.label
            </label>
            <div
              class="container"
            >
              <div
                class="dropbox"
                data-testid="image-dropbox"
                role="presentation"
              >
                <input
                  accept="image/jpg,image/jpeg,image/png"
                  data-testid="file-input"
                  id="image"
                  name="image"
                  style="display: none;"
                  tabindex="-1"
                  type="file"
                />
                <svg>
                  upload.svg
                </svg>
                <p
                  class="typography p promptText"
                >
                  Перетащите изображения сюда
                  <br />
                  или
                  <button
                    class="browseFileLink"
                    type="button"
                  >
                    загрузите
                  </button>
                </p>
                <div
                  class="constraints"
                >
                  <p
                    class="typography p"
                  >
                    imageDropbox.acceptedExtensions
                  </p>
                  <p
                    class="typography p"
                  >
                    imageDropbox.maxFileSize
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="textSection actions"
        >
          <button
            class="actionButton button default"
            id="btn-back"
            type="button"
          >
            formButtons.back
          </button>
          <button
            class="actionButton button primary"
            disabled=""
            id="btn-create"
            type="submit"
          >
            formButtons.create
          </button>
        </div>
      </div>
    </section>
  </form>
</DocumentFragment>
`;
